File: 003/Python/problem003.py
def greatestPrimeFactor(n):
    return gpfSearch(n, 2)
    
def gpfSearch(n, f):
  while n != f:
    if divides(n, f):
      n = n / f
    else:
      f = f + 1
  return f

def divides(n, f):
  return n % f == 0

File: 003/Haskell/Problem003.hs
module Problem003 (greatestPrimeFactor) where

greatestPrimeFactor :: Integer -> Integer
greatestPrimeFactor x = gpfSearch 2 x

gpfSearch f x      | x == f      = x
                   | divides x f = gpfSearch f (div x f)
                   | otherwise   = gpfSearch (f+1) x

divides x y = mod x y == 0